flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/stream/compact/CompactOperator.java

```diff
@@ -0,0 +1,276 @@
+	private final CompactReader.Factory<T> readerFactory;
+	private final CompactWriter.Factory<T> writerFactory;
+
+	private transient FileSystem fileSystem;
+
+	private transient ListState<Map<Long, List<Path>>> expiredFilesState;
+	private transient TreeMap<Long, List<Path>> expiredFiles;
+	private transient List<Path> currentExpiredFiles;
+
+	private transient Set<String> partitions;
+
+	public CompactOperator(
+			SupplierWithException<FileSystem, IOException> fsFactory,
+			CompactReader.Factory<T> readerFactory,
+			CompactWriter.Factory<T> writerFactory) {
+		this.fsFactory = fsFactory;
+		this.readerFactory = readerFactory;
+		this.writerFactory = writerFactory;
+	}
+
+	@Override
+	public void initializeState(StateInitializationContext context) throws Exception {
+		super.initializeState(context);
+		this.partitions = new HashSet<>();
+		this.fileSystem = fsFactory.get();
+
+		ListStateDescriptor<Map<Long, List<Path>>> metaDescriptor =
+				new ListStateDescriptor<>("expired-files", new MapSerializer<>(
+						LongSerializer.INSTANCE,
+						new ListSerializer<>(new KryoSerializer<>(Path.class, getExecutionConfig()))
+				));
+		this.expiredFilesState = context.getOperatorStateStore().getListState(metaDescriptor);
+		this.expiredFiles = new TreeMap<>();
+		this.currentExpiredFiles = new ArrayList<>();
+
+		if (context.isRestored()) {
+			this.expiredFiles.putAll(this.expiredFilesState.get().iterator().next());
+		}
+	}
+
+	@Override
+	public void processElement(StreamRecord<CoordinatorOutput> element) throws Exception {
+		CoordinatorOutput value = element.getValue();
+		if (value instanceof CompactionUnit) {
+			CompactionUnit unit = (CompactionUnit) value;
+			if (unit.isTaskMessage(getRuntimeContext().getNumberOfParallelSubtasks())) {
+				String partition = unit.getPartition();
+				List<Path> paths = unit.getPaths();
+
+				doCompact(paths);
+				this.partitions.add(partition);
+
+				// Only after the current checkpoint is successfully executed can delete
+				// the expired files, so as to ensure the existence of the files.
+				this.currentExpiredFiles.addAll(paths);
+			}
+		} else if (value instanceof EndCompaction) {
+			endCompaction(((EndCompaction) value).getCheckpointId());
+		}
+	}
+
+	private void endCompaction(long checkpoint) {
+		this.output.collect(new StreamRecord<>(new PartitionCommitInfo(
+				checkpoint,
+				getRuntimeContext().getIndexOfThisSubtask(),
+				getRuntimeContext().getNumberOfParallelSubtasks(),
+				new ArrayList<>(this.partitions))));
+		this.partitions.clear();
+	}
+
+	@Override
+	public void snapshotState(StateSnapshotContext context) throws Exception {
+		super.snapshotState(context);
+
+		expiredFilesState.clear();
+		expiredFiles.put(context.getCheckpointId(), new ArrayList<>(currentExpiredFiles));
+		expiredFilesState.add(expiredFiles);
+		currentExpiredFiles.clear();
+	}
+
+	@Override
+	public void notifyCheckpointComplete(long checkpointId) throws Exception {
```

Comment:
If something goes wrong in this method and the job fails over, will this method be called again for the same checkpointId?

Reply:
No, then will wait for next checkpoint notify.